pkg/ingestor/parser/parser.go

```diff
@@ -0,0 +1,261 @@
+		foundBuilders:         []assembler.BuilderNode{},
+		foundDependencies:     []assembler.ArtifactNode{},
+	}
+}
+
+// ParseDocumentTree
+func ParseDocumentTree(docTree processor.DocumentTree) (assembler.AssemblerInput, error) {
+
+	builder := newGraphBuilder()
+	err := builder.parse(docTree)
+	if err != nil {
+		return assembler.AssemblerInput{}, err
+	}
+
+	assemblerinput := assembler.AssemblerInput{
+		Nodes: builder.createNodes(),
+		Edges: builder.createEdges(),
+	}
+
+	return assemblerinput, nil
+}
+
+func (b *graphBuilder) createEdges() []assembler.GuacEdge {
+	edges := []assembler.GuacEdge{}
+	for _, i := range b.foundIdentities {
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.IdentityForEdge{IdentityNode: i, AttestationNode: a})
+		}
+	}
+	for _, s := range b.foundSubjectArtifacts {
+		for _, b := range b.foundBuilders {
+			edges = append(edges, assembler.BuiltByEdge{ArtifactNode: s, BuilderNode: b})
+		}
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.AttestationForEdge{AttestationNode: a, ArtifactNode: s})
+		}
+		for _, d := range b.foundDependencies {
+			edges = append(edges, assembler.DependsOnEdge{ArtifactNode: s, Dependency: d})
+		}
+	}
+	return edges
+}
+
+func (b *graphBuilder) createNodes() []assembler.GuacNode {
+	nodes := []assembler.GuacNode{}
+	for _, i := range b.foundIdentities {
+		nodes = append(nodes, i)
+	}
+	for _, s := range b.foundSubjectArtifacts {
+		nodes = append(nodes, s)
+	}
+	for _, a := range b.foundAttestations {
+		nodes = append(nodes, a)
+	}
+	for _, d := range b.foundDependencies {
+		nodes = append(nodes, d)
+	}
+	for _, b := range b.foundBuilders {
+		nodes = append(nodes, b)
+	}
+	return nodes
+}
+
+func (b *graphBuilder) parse(root processor.DocumentTree) error {
+	err := b.parserHelper(root.Document)
+	if err != nil {
+		return err
+	}
+
+	if len(root.Children) == 0 {
+		return nil
+	}
+
+	for _, c := range root.Children {
+		err := b.parse(c)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (b *graphBuilder) parserHelper(doc *processor.Document) error {
+
+	sub, err := getSubject(doc)
+	if err != nil {
+		return err
+	}
+	b.foundSubjectArtifacts = append(b.foundSubjectArtifacts, sub...)
+
+	dep, err := getDependency(doc)
+	if err != nil {
+		return err
+	}
+	b.foundDependencies = append(b.foundDependencies, dep...)
+
+	id, err := getIdentity(doc)
+	if err != nil {
+		return err
+	}
+	b.foundIdentities = append(b.foundIdentities, id...)
+
+	att, err := getAttestation(doc)
+	if err != nil {
+		return err
+	}
+	b.foundAttestations = append(b.foundAttestations, att...)
+
+	build, err := getBuilder(doc)
+	if err != nil {
+		return err
+	}
+	b.foundBuilders = append(b.foundBuilders, build...)
+
+	return nil
+}
+
+func getSubject(doc *processor.Document) ([]assembler.ArtifactNode, error) {
+	foundSubject := []assembler.ArtifactNode{}
+	switch doc.Type {
+	case processor.DocumentITE6SLSA:
```

Comment:
in the future probably this can be a plugin model as well, for now, would it be possible to put the processing in a function in a `parser_slsa.go` maybe?

Reply:
I will push out another PR for this change. Create an interface and breakout the parser.